lib/SILOptimizer/Mandatory/TFCanonicalizeCFG.cpp

```diff
@@ -233,9 +264,364 @@ SESERegionBuilder::processAcyclicRegionExcludingEnd(SILBasicBlock *startBB,
+  if (changed) {
+    splitAllCondBrCriticalEdgesWithNonTrivialArgs(*F, nullptr, &LI);
+
+    DI.recalculate(*F);
+    PDI.recalculate(*F);
+  }
+}
+
+namespace {
+
+/// Appends the given arguments to the given edge. Deletes the old TermInst
+/// and returns a new TermInst with the appropriate number of args.
+TermInst *appendArguments(TermInst *termInst, SILBasicBlock *target,
+                            ArrayRef<SILValue> newArgs) {
+  SILBuilder builder(termInst);
+  TermInst *newTermInst = nullptr;
+
+  auto createArgsArray = [target,
+                          &newArgs](const OperandValueArrayRef &currentArgs,
+                                    bool appendNewArgs) {
+    SmallVector<SILValue, 8> args(currentArgs.begin(), currentArgs.end());
+    if (appendNewArgs) {
+      args.append(newArgs.begin(), newArgs.end());
+      assert(args.size() == target->getNumArguments() &&
+             "Number of final arguments does not match target's arguments.");
+    }
+    return args;
+  };
+  if (auto *branch = dyn_cast<BranchInst>(termInst)) {
+    assert(branch->getDestBB() == target &&
+           "Incoming edge block and target do not match");
+    SmallVector<SILValue, 8> args =
+        createArgsArray(branch->getArgs(), /*appendNewArgs=*/true);
+    newTermInst =
+        builder.createBranch(branch->getLoc(), branch->getDestBB(), args);
+  } else if (auto *condBranch = dyn_cast<CondBranchInst>(termInst)) {
+    bool isTrueEdge = condBranch->getTrueBB() == target;
+    assert(((isTrueEdge && condBranch->getTrueBB() == target) ||
+            (!isTrueEdge && condBranch->getFalseBB() == target)) &&
+           "Incoming edge block and target do not match");
+    SmallVector<SILValue, 8> trueArgs = createArgsArray(
+        condBranch->getTrueArgs(), /*appendNewArgs=*/isTrueEdge);
+    SmallVector<SILValue, 8> falseArgs = createArgsArray(
+        condBranch->getFalseArgs(), /*appendNewArgs=*/!isTrueEdge);
+    newTermInst = builder.createCondBranch(
+        condBranch->getLoc(), condBranch->getCondition(),
+        condBranch->getTrueBB(), trueArgs, condBranch->getFalseBB(), falseArgs,
+        condBranch->getTrueBBCount(), condBranch->getFalseBBCount());
+  } else {
+    // At the moment we can only add arguments to br and cond_br.
+    llvm_unreachable("Can't add argument to terminator");
+  }
+  // Remove the old terminator instruction.
+  termInst->dropAllReferences();
+  termInst->eraseFromParent();
+  return newTermInst;
+}
+
+/// Creates a TF handle of appropriate type that is initialized to the given
+/// element value.
+SILValue createTFHandleFromElement(SILBuilder &builder, SILLocation location,
+                                   SILValue element) {
+  assert(builder.hasValidInsertionPoint());
+  ASTContext& context = builder.getASTContext();
+  SILType elementType = element->getType();
+  // Literals take attributes specifying the dtype, value, and device.
+  std::string opName("__tfop_Const,dtype$dtype,value$tensor,__device");
```

Comment:
The `graph_op` name should be `Const`, not the "__tfop_" encoding that builtins are using. Dtypes are already implied when you specify operands. `__device` is an attribute in this instruction.

Reply:
See usage here: https://github.com/apple/swift/blob/39ced43dd4c59095fe0b7135ccbf42098fd7b410/test/TensorFlow/integration.swift#L64

Reply:
Thanks for the pointer. I think I am not setting the attributes properly as well. Is there a code snippet in C++ where "Const" graph operation is created?

Reply:
https://github.com/apple/swift/blob/39ced43dd4c59095fe0b7135ccbf42098fd7b410/lib/SILOptimizer/Mandatory/TFPartition.cpp#L2977

Reply:
Thanks for the pointer. I changed all the relevant builtin instructions to graph operations. PTAL.